### PR TITLE
fix safety of ChooseResult.save

### PR DESCRIPTION
**Pull request by aG0aep6G**
Hopefully, this can unstuck the stable merge #6943.

The fix isn't as pretty as I'd like, but it's the best I could find. The difficulty is in trusting only the access to the union. Copies and `.save` calls can't be trusted.

In #6904, Walter solved this by adding `@trusted` helpers like this:

    ref get1(return ref ChooseResult r) @trusted { return r.r1; }

That would work here too, but I think it's technically a violation of the `@trusted` contract. That function does not have a safe interface.

### CODE REVIEW

**Comment by aG0aep6G** [comment]
> There is an issue on bugzilla for please reference it in the commit message.

I don't think there is a Bugzilla issue for this.

**Comment by wilzbach** [comment]
CircleCI failure can be fixed by rebasing to the latest upstream/master.

**Comment by aG0aep6G** [comment]
> CircleCI failure can be fixed by rebasing to the latest upstream/master.

This is against stable.

File: std/range/package.d [hunk 1]
```diff
@@ -1626,6 +1653,78 @@ pure @safe unittest // issue 18657
...
+        void popFront() {}
+        @property R save() { p = q; return this; }
+            // `p = q;` is only there to prevent inference of `scope return`.
+    }
+    R r;
+    choose(true, r, r).save;
+}
+
+// Make sure ChooseResult.save doesn't trust @system user code.
+@system unittest // copy is @system
```
Comment (wilzbach) [inline std/range/package.d]: Shouldn't this have a `__traits(compiles` with @safe in there too ensure that this never gets safe?

Comment (aG0aep6G) [reply std/range/package.d]: That's the next unittest.

**Comment by wilzbach** [comment]
Hmm, the Vibe.d failure is still there:

```
mongodb/.dub/build/vibe-d-mongodb-test-library-unittest-linux.posix-x86_64-dmd_2085-36D85DDF181792A215C97F8B95F2BD77/vibe-d-mongodb-test-library.o:../.dub/packages/vibe-core-1.6.3-alpha.1/vibe-core/source/vibe/appmain.d:TypeInfo_S3std8typecons__T5TupleTiTdZQl.__init: error: undefined reference to 'std.typecons.Tuple!(int, double).Tuple.__init'
mongodb/.dub/build/vibe-d-mongodb-test-library-unittest-linux.posix-x86_64-dmd_2085-36D85DDF181792A215C97F8B95F2BD77/vibe-d-mongodb-test-library.o:../.dub/packages/vibe-core-1.6.3-alpha.1/vibe-core/source/vibe/appmain.d:function bool std.typecons.Tuple!(int, double).Tuple.__xopEquals(ref const(std.typecons.Tuple!(int, double).Tuple), ref const(std.typecons.Tuple!(int, double).Tuple)): error: undefined reference to 'const pure nothrow @nogc @safe bool std.typecons.Tuple!(int, double).Tuple.opEquals!(const(std.typecons.Tuple!(int, double).Tuple)).opEquals(const(std.typecons.Tuple!(int, double).Tuple))'
mongodb/.dub/build/vibe-d-mongodb-test-library-unittest-linux.posix-x86_64-dmd_2085-36D85DDF181792A215C97F8B95F2BD77/vibe-d-mongodb-test-library.o:../.dub/packages/vibe-core-1.6.3-alpha.1/vibe-core/source/vibe/appmain.d:function int std.typecons.Tuple!(int, double).Tuple.__xopCmp(ref const(std.typecons.Tuple!(int, double).Tuple), ref const(std.typecons.Tuple!(int, double).Tuple)): error: undefined reference to 'const pure nothrow @nogc @safe int std.typecons.Tuple!(int, double).Tuple.opCmp!(const(std.typecons.Tuple!(int, double).Tuple)).opCmp(const(std.typecons.Tuple!(int, double).Tuple))'
```

but apparently that affects only upstream/stable and appeared in others too:

https://buildkite.com/dlang/phobos/builds/1285

**Comment by wilzbach** [comment]
Thanks a lot for this!
FYI: I opened a new stable->master PR with this:

https://github.com/dlang/phobos/pull/6951

**Pull request merged**